pkg/kyverno/test/command.go

```diff
@@ -0,0 +1,536 @@
+				delete(policy.Annotations, engine.PodControllersAnnotation)
+			}
+		}
+	}
+	policyCtx := &engine.PolicyContext{Policy: *policy, NewResource: mutateResponse.PatchedResource, JSONContext: ctx}
+	validateResponse := engine.Validate(policyCtx)
+	var policyHasGenerate bool
+	for _, rule := range policy.Spec.Rules {
+		if rule.HasGenerate() {
+			policyHasGenerate = true
+		}
+	}
+	if policyHasGenerate {
+		generateResponse := engine.Generate(engine.PolicyContext{Policy: *policy, NewResource: *resource})
+		engineResponses = append(engineResponses, generateResponse)
+		if len(generateResponse.PolicyResponse.Rules) > 0 {
+			log.Log.V(3).Info("generate resource is valid", "policy", policy.Name, "resource", resPath)
+		} else {
+			fmt.Printf("generate policy %s resource %s is invalid \n", policy.Name, resPath)
+			for i, r := range generateResponse.PolicyResponse.Rules {
+				fmt.Printf("%d. %s \b", i+1, r.Message)
+			}
+
+		}
+	}
+	return engineResponses, validateResponse, nil
+}
+
+func buildPolicyResults(resps []*response.EngineResponse) map[string][]interface{} {
+	results := make(map[string][]interface{})
+	infos := policyreport.GeneratePRsFromEngineResponse(resps, log.Log)
+	for _, info := range infos {
+		for _, infoResult := range info.Results {
+			for _, rule := range infoResult.Rules {
+				if rule.Type != utils.Validation.String() {
+					continue
+				}
+				result := report.PolicyReportResult{
+					Policy: info.PolicyName,
+					Resources: []*corev1.ObjectReference{
+						{
+							Name:       infoResult.Resource.Name,
+						},
+					},	
+				}
+				result.Rule = rule.Name
+				result.Status = report.PolicyStatus(rule.Check)
+				results[rule.Name] = append(results[rule.Name], result)
+			}
+		}
+	}
+	return results
+}
+func getVariable( valuesFile string) ( valuesMap map[string]map[string]Resource, err error) {
+	if valuesFile != "" {
+		yamlFile, err := ioutil.ReadFile(valuesFile)
+		if err != nil {
+			return valuesMap, sanitizederror.NewWithError("unable to read yaml", err)
+		}
+		valuesBytes, err := yaml.ToJSON(yamlFile)
+		if err != nil {
+			return valuesMap, sanitizederror.NewWithError("failed to convert json", err)
+		}
+		values := &Values{}
+		if err := json.Unmarshal(valuesBytes, values); err != nil {
+			return valuesMap, sanitizederror.NewWithError("failed to decode yaml", err)
+		}
+		for _, p := range values.Policies {
+			pmap := make(map[string]Resource)
+			for _, r := range p.Resources {
+				pmap[r.Name] = r
+			}
+			valuesMap[p.Name] = pmap
+		}
+	}
+	return  valuesMap, nil
+}
+
+func removeDuplicatevariables(matches [][]string) string {
+	var variableStr string
+	for _, m := range matches {
+		for _, v := range m {
+			foundVariable := strings.Contains(variableStr, v)
+			if !foundVariable {
+				variableStr = variableStr + " " + v
+			}
+		}
+	}
+	return variableStr
+}
+
+func applyPoliciesFromPath(fs billy.Filesystem, policyBytes []byte, valuesFile string, isGit bool) (err error) {
```

Comment:
can we not reuse this logic from apply? 

if the difference here is the FS, lets extract that into a type that can be used in both places.

Reply:
Yes, added in common pkg. using for both command test/apply